### PR TITLE
VIH-9765 filter non availability hours by future dates 

### PR DESCRIPTION
**Pull request by abbasiTahir**
**Before creating a pull request make sure that:**

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/VIH-9765

### Change description ###

Applied the filter to display data for future dates , fixed the broken unit tests due to change . Added unit test to cover the code .

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

**Review by will-craig — CHANGES_REQUESTED** [review]
Sorry these changes to the tests arn't viable. They may work now temporarily. But pushing the date back to get them to work means once we reach October/November of this year they will start failing again. You'll need to rethink this.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.spec.ts [hunk 1]
```diff
@@ -986,6 +986,22 @@ describe('VhoNonAvailabilityWorkHoursTableComponent', () => {
...
+
+        it('should on return future dates vho data ', () => {
+            // arrange
+          const futureDates =  component.nonAvailabilityWorkHoursResponses = [
+                new VhoNonAvailabilityWorkHoursResponse({ id: 0, start_time: new Date('2023/03/03'), end_time: new Date('2023/03/04') }),
+                new VhoNonAvailabilityWorkHoursResponse({ id: 1, start_time: new Date('2023/03/06'), end_time: new Date('2023/03/07') }),
+                new VhoNonAvailabilityWorkHoursResponse({ id: 2, start_time: new Date('2023/03/09'), end_time: new Date('2023/03/12') })
+            ];
+            fixture.detectChanges();
+
```
Comment (will-craig) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.spec.ts:998]: todaysDate = new Date()

File: AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts [hunk 2]
```diff
@@ -453,5 +455,10 @@ export class VhoWorkHoursNonAvailabilityTableComponent implements OnInit, CanDea
         } else if (this.nonWorkHours.length === 0) {
             this.showMessage(VhoWorkHoursNonAvailabilityTableComponent.WarningNoWorkingHoursForVho);
         }
+
+    }
+    public filterByFutureDate(value: VhoNonAvailabilityWorkHoursResponse[]) {
+        const todayDate = new Date();
```
Comment (will-craig) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts:461]: Make todays data a property of the component, then we can test this without updating the other unit tests

Comment (will-craig) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts:461]:
```suggestion
        const todayDate =  this.todaysDate;
```

Comment (shaed-parkar) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts:461]: I agree, we should make the date property flexible else the tests will fail next week

File: AdminWebsite/AdminWebsite/ClientApp/src/app/testing/data/work-allocation-test-data.ts [hunk 3]
```diff
@@ -6,108 +6,108 @@ export class MockWorkAllocationValues {
          [
             {
                "id":1,
-               "end_time":"2023-01-01T15:00:00Z",
```
Comment (shaed-parkar) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/testing/data/work-allocation-test-data.ts:9]: Please revert this change. After you implement @will-craig 's suggestion to use a filter variable you can adjust the test setup to minimise changes

File: AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts [hunk 4]
```diff
@@ -44,6 +44,7 @@ export class VhoWorkHoursNonAvailabilityTableComponent implements OnInit, CanDea
         }
         this.hideMessage();
         if (value) {
+           value = this.filterByFutureDate(value);
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts:47]: I think that instead of passing all the data to the table component, just to filter it out, it would be better to filter it early and only pass data to the table that it should actually display. Maybe keep this table as a simple, presentation component.

Comment (shaed-parkar) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts:47]: I would have to agree here too. For a table component that I would expect to traditionally be a presentation component, it's doing a fair bit. If possible, I would suggest you filter the data before passing it into the table component.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts (hunk 2 above)
Comment (abbasiTahir) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts]: applied the suggested change .

Comment (abbasiTahir) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts]: applied the suggested changed .

**Review by will-craig — DISMISSED** [review]

File: AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts (hunk 4 above)
Comment (will-craig) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts]: I suggested Tahir do it here as the search component also fetches data for other tables, and as the filter is only required on this specific table, thought the responsibility should live there. Understand your reasonings however, makes sense, so I'm not too bothered when the logic sits.

Comment (abbasiTahir) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts]: Addressed the issue mentioned in comment .

File: AdminWebsite/AdminWebsite/ClientApp/src/app/testing/data/work-allocation-test-data.ts (hunk 3 above)
Comment (abbasiTahir) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/testing/data/work-allocation-test-data.ts]: done

File: AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts (hunk 4 above)
Comment (shaed-parkar) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts]: I just realised the result field is used by two different components which both accept a different type. Ideally they should be two separate fields but that might have to be saved for another day

**Review by shaed-parkar — APPROVED** [review]

**Review by shaed-parkar — APPROVED** [review]
Would be good to stop the result field being shared across two different components

File: AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts (hunk 4 above)
Comment (abbasiTahir) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/edit-work-hours/vho-work-hours-non-availability-table/vho-work-hours-non-availability-table.component.ts]: agreed shaed

**Review by phillipjchambers — APPROVED** [review]

**Review by will-craig — APPROVED** [review]
